src/System.Net.NetworkInformation/src/System/Net/NetworkInformation/UnixNetworkInterface.cs

```diff
@@ -0,0 +1,153 @@
+
+                case Interop.Sys.ARPHRD_PRONET:
+                    return NetworkInterfaceType.TokenRing;
+
+                case Interop.Sys.ARPHRD_ATM:
+                    return NetworkInterfaceType.Atm;
+
+                case Interop.Sys.ARPHRD_SLIP:
+                case Interop.Sys.ARPHRD_CSLIP:
+                case Interop.Sys.ARPHRD_SLIP6:
+                case Interop.Sys.ARPHRD_CSLIP6:
+                    return NetworkInterfaceType.Slip;
+
+                case Interop.Sys.ARPHRD_PPP:
+                    return NetworkInterfaceType.Ppp;
+
+                case Interop.Sys.ARPHRD_TUNNEL:
+                case Interop.Sys.ARPHRD_TUNNEL6:
+                    return NetworkInterfaceType.Tunnel;
+
+                case Interop.Sys.ARPHRD_LOOPBACK:
+                    return NetworkInterfaceType.Loopback;
+
+                case Interop.Sys.ARPHRD_FDDI:
+                    return NetworkInterfaceType.Fddi;
+
+                case Interop.Sys.ARPHRD_IEEE80211:
+                case Interop.Sys.ARPHRD_IEEE80211_PRISM:
+                case Interop.Sys.ARPHRD_IEEE80211_RADIOTAP:
+                    return NetworkInterfaceType.Wireless80211;
+
+                default:
+                    Debug.WriteLine("Unmapped ARP Hardware type: " + arpHardwareType);
+                    return NetworkInterfaceType.Unknown;
+            }
+        }
+
+        public override string Name { get { return _name; } }
+
+        public override NetworkInterfaceType NetworkInterfaceType { get { return _networkInterfaceType; } }
+
+        public override PhysicalAddress GetPhysicalAddress()
+        {
+            Debug.Assert(_physicalAddress != null, "_physicalAddress was never initialized. This means no address with type AF_PACKET was discovered.");
+            return _physicalAddress;
+        }
+
+        /// <summary>
+        /// The system's index for this network device.
+        /// </summary>
+        public int Index { get { return _index; } }
+
+        /// <summary>
+        /// Returns a list of all of the interface's IP Addresses.
+        /// </summary>
+        public List<IPAddress> Addresses { get { return _addresses; } }
+
+        // Adds any IPAddress to this interface's List of addresses.
+        protected void AddAddress(IPAddress ipAddress)
+        {
+            _addresses.Add(ipAddress);
+        }
+
+        public IPAddress GetNetMaskForIPv4Address(IPAddress address)
+        {
+            Debug.Assert(address.AddressFamily == Sockets.AddressFamily.InterNetwork);
+            return _netMasks[address];
+        }
+
+        protected static unsafe void ProcessIpv4Address(UnixNetworkInterface uni,
+                                                Interop.Sys.IpAddressInfo* addressInfo,
+                                                Interop.Sys.IpAddressInfo* netMask)
+        {
+            byte[] ipBytes = new byte[addressInfo->NumAddressBytes];
+            fixed (byte* ipArrayPtr = ipBytes)
```

Comment:
When I have time to refactor for perf improvements I will go through and check all my array copying/marshalling.